test/unit/test-mysqlDao.js

```diff
@@ -231,6 +258,104 @@ describe('MysqlDao', function() {
         });
     });
 
+    describe('upsertBulk', function() {
+        beforeEach(databaseSetup);
+
+        it('Should create multiple rows in the top_secret_info table', function() {
+            const insertArgs = [
+                {
+                    secret_agent_code: 1, handler_code: 1, fake_passport_number: 1, bank_account: 1,
+                    code_name: 'Duchess', intel: 'The Briefcase'
+                },
+                {
+                    secret_agent_code: 2, handler_code: 2, fake_passport_number: 2, bank_account: 2,
+                    code_name: 'Otter', intel: 'Dead Drop'
+                },
+                {
+                    secret_agent_code: 3, handler_code: 3, fake_passport_number: 3, bank_account: 3,
+                    code_name: 'Pinto', intel: 'Poison Pill'
+                }
+            ];
+            const onDupUpdateArgs = ['intel'];
+
+            return topSecretInfoDao.upsertBulk(insertArgs, onDupUpdateArgs)
+                .then(result => {
+                    Should.exist(result);
+                    result.should.eql(3);
+                });
+        });
+
+        it('Should update a row with a value not specified in the insert', function() {
+            const insertArgs = [
+                {
+                    secret_agent_code: 1, handler_code: 1, fake_passport_number: 1, bank_account: 1,
+                    code_name: 'Duchess', intel: 'The Briefcase'
+                }
+            ];
+
+            const onDupUpdateArgs = ['intel'];
+
+            return topSecretInfoDao.upsertBulk(insertArgs, onDupUpdateArgs)
+                .then(result => {
+                    let promise = topSecretInfoDao.upsertBulk(insertArgs,  { intel: 'Troop Movements' });
+                    return Promise.all([promise]).then(results => results[0]);
+                })
+                .then(result => {
+                    Should.exist(result);
+                    result.should.eql(2);
+
+                    const assertPromise = topSecretInfoDao.getOne({secretAgentCode: 1})
+                        .then(info => {
+                            Should.exist(info);
+                            info.intel.should.eql('Troop Movements');
+                        });
+                    return Promise.all([assertPromise]);
```

Comment:
Could be a personal preference, but I like to do all mutating/fetching in the first few promises, and then do all asserting after everything has been fetched and saved.

For example: 
```javascript
            let upsertResult;
            return topSecretInfoDao.upsertBulk(insertArgs, onDupUpdateArgs)
                .then(() => {
                    return topSecretInfoDao.upsertBulk(insertArgs,  { intel: 'Troop Movements' });
                })
                .then(results => {
                    upsertResult = results;
                    return topSecretInfoDao.getOne({secretAgentCode: 1});
                })
                .then(topSecretInfo => {
                    Should.exist(topSecretInfo);
                    topSecretInfo.intel.should.eql('Troop Movements');

                    Should.exist(upsertResult);
                    upsertResult.should.eql(2);
                });
```